QueryCiv3/Bic.cs

```diff
@@ -9,9 +10,253 @@ public class BicData
         public bool HasCustomRules => Bic.SectionExists("BLDG");
         public bool HasCustomMap => Bic.SectionExists("WCHR");
 
-        public BicData(byte[] bicBytes)
+        public BLDG[] Bldg;
+        public CITY[] City;
+        public CLNY[] Clny;
+        public CONT[] Cont;
+        public CTZN[] Ctzn;
+        public CULT[] Cult;
+        public DIFF[] Diff;
+        public ERAS[] Eras;
+        public ESPN[] Espn;
+        public EXPR[] Expr;
+        public FLAV[] Flav;
+        public GAME[] Game;
+        public GOOD[] Good;
+        public GOVT[] Govt;
+        public LEAD[] Lead;
+        public PRTO[] Prto;
+        public RACE[] Race;
+        public RULE[] Rule;
+        public SLOC[] Sloc;
+        public TECH[] Tech;
+        public TERR[] Terr;
+        public TFRM[] Tfrm;
+        public TILE[] Tile;
+        public UNIT[] Unit;
+        public WCHR[] Wchr;
+        public WMAP[] Wmap;
+        public WSIZ[] Wsiz;
+
+        public bool[,] TerrGood; // which resources are allowed on which types of terrain
+        public GOVTGOVT[,] GovtGovt; // relationships between governments
+
+        public unsafe BicData(byte[] bicBytes)
         {
             Bic = new Civ3File(bicBytes);
+
+            fixed (byte* bytePtr = bicBytes)
+            {
+                int offset = 736; // byte index of the first header in BIQ files
+                string header;
+                int count = 0;
+                int dataLength = 0;
+
+                while (offset < bicBytes.Length - 12) { // Don't read past the end
```

Comment:
The -12 is that so when the loop is in linear-search mode, the offset isn't overflown(?) when trying to get the next potential header and count:
`header = Bic.GetString(offset, 4);
count = Bic.ReadInt32(offset + 4);`
Really, it should be -8, since those together add up to only 8, but either way, it should only be a temporary thing. The linear-search nature of the loop will only be there until the rest of the BIQ sections are defined, at which point the loop should be able to jump from one header to the next with no searching necessary. Then we can change the default case to throwing an exception, because the default would only occur in situations where something went wrong with the memory reading (bad file).

Reply:
That makes more sense with the 8, and I agree that the check (in its current form, at least) will be temporary.  We might still need a check, but it could be a check against the whole length (assuming the offset winds up being equal to the whole length after the last section, whichever section that is).